### DISCUSSION TITLE
Converted deconvolution data into spike train (binary point)

### DISCUSSION THREAD

**Discussion by zhouyi0812** [flatironinstitute/CaImAn#1195] [Q&A]
Hello!

After I download the deconvolution trace from estimates.S, I got deconvoluted time series data. However, I want the spike train (point data) so I use another package to select the peak time point and convert it to 1 and the rest of the time point is converted to 0. I wonder are you also doing this way?  If there is another way to handle it in CaImAn that I missed.

Thank you!!!

**Comment by kushalkolar** [comment]
Do you just want to do 0 - 1 normalization? The easiest thing is to just use tslearn min max scaler: https://tslearn.readthedocs.io/en/stable/gen_modules/preprocessing/tslearn.preprocessing.TimeSeriesScalerMinMax.html

**Comment by zhouyi0812** [comment]
<img width="501" alt="Screen Shot 2023-10-17 at 5 32 22 PM" src="https://github.com/flatironinstitute/CaImAn/assets/48576330/de9b7c10-f590-4841-8fa1-a284d1576b83">

I think it is not normalized to (0,1). For example, the deconvolution trace is 0, 0, 4.3, 5.4, 6.6, 0, 0 and 6.6 is the peak and my final data is 0, 0, 0, 0, 1, 0, 0

**Reply by kushalkolar** [reply]
So you just want an array where the max is 1 and rest 0? Not sure what you're really trying to do but:

```python
a[a < a.max()] = 0
a[a > 0] = 1
```

**Comment by zhouyi0812** [comment]
You're not doing this way? So how you plot this kind of raster plot?
<img width="722" alt="Screen Shot 2023-10-11 at 6 08 59 PM" src="https://github.com/flatironinstitute/CaImAn/assets/48576330/cad8a4d2-51f3-4b42-a74e-da248e26ed13">

**Comment by zhouyi0812** [comment]
![Uploading Screen Shot 2023-10-17 at 6.18.24 PM.png…]()

**Comment by zhouyi0812** [comment]
<img width="467" alt="Screen Shot 2023-10-17 at 6 20 32 PM" src="https://github.com/flatironinstitute/CaImAn/assets/48576330/c86cad04-2456-4c4a-a7ff-a1a8de84937b">

**Reply by kushalkolar** [reply]
Converting S to spike count isn't as trivial as max(S). See this discussion: https://github.com/flatironinstitute/CaImAn/issues/499

**Comment by EricThomson** [comment]
I interpreted it as you want a raster plot like thing. This is not currently part of the `estimates` object, which gives you an estimate of spike count in `estimates.S` on the same time scale as the sampling rate of your experiment.  It is not possible, afaik to get a finer time scale. This is using OASIS (implemented in `constrained_oasisAR2()` function), and things get more complex with other algorithms like `foopsi`. OASIS works really well so I stick with that. 

Though you didn't say what type of data you are analyzing iirc you are looking at 1p. If 1p data I would be wary of deconvolution results tbh we don't have any ground truth data.  I'd probably stick with analyzing traces if using CNMFE.

**Reply by zhouyi0812** [reply]
So basically I need to get the trace from here: https://github.com/EricThomson/yacare/blob/main/normalize_1p_traces.ipynb

Then find the peak and plot the raster plot. Thank you!

**Reply by kushalkolar** [reply]
It still wouldn't make much sense to set the max as 1 and everything else as 0 if you want to understand the activity of that neuron. You can plot a heatmap of dfof or zscored activity using seaborn, for example with C:

```python
import seaborn as sns
from tslearn.preprocessing import TimeSeriesScalerMeanVariance

zscore = TimeSeriesScalerMeanVariance().fit_transform(cnmf_obj.estimates.C)

sns.heatmap(zscore)
```

**Reply by zhouyi0812** [reply]
Thank you very much! I want to use that because I saw some downstream analysis from this tool: https://elephant.readthedocs.io/en/stable/ 

Also, may I ask have you used multivariate granger causality to analyze and cluster the trace? Thank you!

**Comment by kushalkolar** [comment]
Let me try to clear one misconception, you can't really deconvolve calcium
traces into ephys waveforms. You can get rough estimates of spike rate.
This paper has a good into that discusses it:
https://www.nature.com/articles/s41593-021-00895-5

**Reply by zhouyi0812** [reply]
Thank you very much for your clarification! 
While I trying to use the code above, it seems I need to save hdf5 file and then convert to cnmf.object 
I am confused about this 'batch_dataframe_file.pickle' and replace with my hdf5 file. But it says AttributeError: 'CNMF' object has no attribute 'iloc'. Could you give some guidance? Thank you!

from mesmerize_core import load_batch

df = load_batch("/path/to/batch_dataframe_file.pickle")

# assume the 0th index is a cnmf item
cnmf_obj = df.iloc[0].cnmf.get_output()

# see some estimates
print(cnmf_obj.estimates.C)
print(cnmf_obj.estimates.f)

**Reply by EricThomson** [reply]
@zhouyi0812 it sounds like this last question is a mesmerize question, have you resolved it?
